src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,287 @@
+                // Type.fieldName
+                // if instance field we will use the same thechnique as in invokeOpToTree
+                int flags;
+                try {
+                    flags = fieldLoadOp.fieldDescriptor().resolveToMember(MethodHandles.lookup()).getModifiers();
+                } catch (ReflectiveOperationException e) {
+                    throw new RuntimeException(e);
+                }
+                var name = names.fromString(fieldLoadOp.fieldDescriptor().name());
+                var type = typeElementToType(fieldLoadOp.resultType());
+                var owner = typeElementToType(fieldLoadOp.fieldDescriptor().refType());
+                var sym = new Symbol.VarSymbol(flags, name, type, owner.tsym);
+                yield treeMaker.Select(treeMaker.Ident(owner.tsym), sym);
+            }
+            case CoreOp.ArrayAccessOp.ArrayStoreOp arrayStoreOp -> {
+                var array = arrayStoreOp.operands().get(0);
+                var val = arrayStoreOp.operands().get(1);
+                var index = arrayStoreOp.operands().get(2);
+                var as = treeMaker.Assign(
+                        treeMaker.Indexed((JCTree.JCExpression) valueToTree.get(array), (JCTree.JCExpression) valueToTree.get(index)),
+                        (JCTree.JCExpression) valueToTree.get(val)
+                );
+                yield treeMaker.Exec(as);
+                // body builder are created but never passed when creating the op, why ?
+            }
+            default -> throw new IllegalStateException("Op -> JCTree not supported for :" + op.getClass().getName());
+        };
+        valueToTree.put(op.result(), tree);
+        return tree;
+    }
+
+    public JCTree.JCMethodDecl transformFuncOpToAST(CoreOp.FuncOp funcOp, Name methodName) {
+        Assert.check(funcOp.body().blocks().size() == 1);
+
+        var paramTypes = List.of(crSym.opFactoryType, crSym.typeElementFactoryType);
+        var mt = new Type.MethodType(paramTypes, crSym.opType, List.nil(), syms.methodClass);
+        ms = new Symbol.MethodSymbol(PUBLIC | STATIC | SYNTHETIC, methodName, mt, currClassSym);
+        currClassSym.members().enter(ms);
+
+        // TODO add VarOps in OpBuilder
+        funcOp = addVarsWhenNecessary(funcOp);
+        funcOp.writeTo(System.out);
+        for (int i = 0; i < funcOp.parameters().size(); i++) {
+            valueToTree.put(funcOp.parameters().get(i), treeMaker.Ident(ms.params().get(i)));
+        }
+
+        var stats = new ListBuffer<JCTree.JCStatement>();
+        for (Op op : funcOp.body().entryBlock().ops()) {
+            var tree = opToTree(op);
+            if (tree instanceof JCTree.JCStatement stat) {
+                stats.add(stat);
+            }
+        }
+        var mb = treeMaker.Block(0, stats.toList());
+
+        return treeMaker.MethodDef(ms, mb);
+    }
+
+    public static CoreOp.FuncOp addVarsWhenNecessary(CoreOp.FuncOp funcOp) {
```

Comment:
Maybe I'm missing something obvious. But if you have something like:
```
func @"f" ()java.util.Map -> {
    %0 : java.util.Map<java.lang.Integer, java.lang.Integer> = new @"func<java.util.HashMap>";
    %1 : int = constant @"1";
    %2 : int = constant @"2";
    %3 : java.lang.Object = invoke %0 %1 %2 @"java.util.Map::put(java.lang.Object, java.lang.Object)java.lang.Object";
    return %0;
}
```

Can't we translate as:

```
Map $0 = new HashMap();
int $1 = 1;
int $2 = 3;    
Object $3 = $0.put($1, $2);
return $0;
```

Note how this translation is very 1-1 and direct, and doesn't require any adaptation step in the model. We just emit a new local variable for each op we see in the body -- and then refer to operands via their corresponding (locally declared) variable.

Reply:
(of course we'd only do the above translation for ops that have a non-void result type)

Reply:
We can follow up on this improved translation at a later point.

Reply:
Now I see. The current approach is inserting vars only when required. You're saying we can insert Vars for all operation (which is a simpler approach) ?